zrml/asset-router/src/lib.rs

```diff
@@ -0,0 +1,742 @@
+            };
+
+            let by_balance = TryInto::<Self::Balance>::try_into(by_amount_abs)
+                .map_err(|_| Error::<T>::AmountIntoBalanceFailed)?;
+            if by_amount.is_positive() {
+                Self::deposit(currency_id, who, by_balance)
+            } else {
+                Self::withdraw(currency_id, who, by_balance).map(|_| ())
+            }
+        }
+    }
+
+    impl<T: Config> MultiLockableCurrency<T::AccountId> for Pallet<T> {
+        type Moment = T::BlockNumber;
+
+        fn set_lock(
+            lock_id: LockIdentifier,
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> DispatchResult {
+            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                return <T::Currencies as MultiLockableCurrency<T::AccountId>>::set_lock(
+                    lock_id, currency, who, amount,
+                );
+            }
+
+            Err(Error::<T>::Unsupported.into())
+        }
+
+        fn extend_lock(
+            lock_id: LockIdentifier,
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> DispatchResult {
+            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                return <T::Currencies as MultiLockableCurrency<T::AccountId>>::extend_lock(
+                    lock_id, currency, who, amount,
+                );
+            }
+
+            Err(Error::<T>::Unsupported.into())
+        }
+
+        fn remove_lock(
+            lock_id: LockIdentifier,
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+        ) -> DispatchResult {
+            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                return <T::Currencies as MultiLockableCurrency<T::AccountId>>::remove_lock(
+                    lock_id, currency, who,
+                );
+            }
+
+            Err(Error::<T>::Unsupported.into())
+        }
+    }
+
+    impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
+        fn can_reserve(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            value: Self::Balance,
+        ) -> bool {
+            only_currency!(currency_id, false, MultiReservableCurrency, can_reserve, who, value)
+        }
+
+        fn slash_reserved(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            value: Self::Balance,
+        ) -> Self::Balance {
+            only_currency!(currency_id, value, MultiReservableCurrency, slash_reserved, who, value)
+        }
+
+        fn reserved_balance(currency_id: Self::CurrencyId, who: &T::AccountId) -> Self::Balance {
+            only_currency!(
+                currency_id,
+                Zero::zero(),
+                MultiReservableCurrency,
+                reserved_balance,
+                who
+            )
+        }
```

Comment:
Silently suppressing the error from `only_currency` seems concerning. Sure, this is not called directly by user, rather by us developers. I am suggesting to log an error and using `debug_assert` to fail for tests.

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1189/commits/eeb43e36a666a8be3bde65f0000f78cabb1f43ce